openmls/src/binary_tree/array_representation/treemath.rs

```diff
@@ -43,106 +151,113 @@ fn level(index: NodeIndex) -> usize {
     k
 }
 
-pub(super) fn root(size: NodeIndex) -> NodeIndex {
-    (1 << log2(size)) - 1
+pub(super) fn root(size: u32) -> TreeNodeIndex {
+    debug_assert!(size > 0);
+    TreeNodeIndex::new((1 << log2(size)) - 1)
 }
 
-pub(super) fn left(index: NodeIndex) -> Result<NodeIndex, TreeMathError> {
-    let x = index;
+pub(super) fn left(index: ParentNodeIndex) -> TreeNodeIndex {
+    let x = index.to_tree_index();
     let k = level(x);
-    if k == 0 {
-        return Err(TreeMathError::LeafHasNoChildren);
-    }
-    Ok(x ^ (0x01 << (k - 1)))
+    debug_assert!(k > 0);
+    let index = x ^ (0x01 << (k - 1));
+    TreeNodeIndex::new(index)
 }
 
-pub(super) fn right(index: NodeIndex, size: NodeIndex) -> Result<NodeIndex, TreeMathError> {
-    let x = index;
-    let n = size;
+fn left_helper(x: u32) -> u32 {
     let k = level(x);
-    if k == 0 {
-        return Err(TreeMathError::LeafHasNoChildren);
-    }
+    debug_assert!(k > 0);
+    x ^ (0x01 << (k - 1))
+}
+
+pub(super) fn right(index: ParentNodeIndex, size: u32) -> TreeNodeIndex {
+    debug_assert!(size > 0);
+    debug_assert!(index.to_tree_index() < size);
+    let x = index.to_tree_index();
+    let k = level(x);
+    debug_assert!(k > 0);
```

Comment:
I understand how the level of k has to be > 0 due to the implementation of `ParentNodeIndex`, but `size` has no protection at all.

Reply:
I introduced `TreeSize` now.